### PR TITLE
Initialize database from snapshot

### PR DESCRIPTION
**Pull request by jmederosalvarado**
## Changes

- Adds support for initializing the client db from a snapshot file in zip format

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

## Testing

It's a very simple plugin, which is disabled by default, no risk of introducing regression. I tested the functionality using Sepolia.

#### Requires testing

- [ ] Yes
- [x] No

#### Requires documentation update

- [x] Yes
- [ ] No

Documentation Issue: #6178

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

_If yes, fill in the details here. Remove if not applicable._

### CODE REVIEW

**Comment by jmederosalvarado** [comment]
> Progress tracker does not seem to be tracking download progress. I would track that as well

Why do you say it's not tracking download progress? reading from the response stream is what progresses the download, and it's being tracked and logged. See [here](https://github.com/NethermindEth/nethermind/pull/6177/files#diff-9d81e43b7817ba0d761c63003a8bfcba00705c6e4059cadd793e8cd32965f76eR110-R122)

**Comment by smartprogrammer93** [comment]
Sorry, you are right. It is picking the download progress but not the extract progress. Is there a way to add that?

**Review by smartprogrammer93 — APPROVED** [review]

**Comment by jmederosalvarado** [comment]
> Sorry, you are right. It is picking the download progress but not the extract progress. Is there a way to add that?

No, I have no idea how to do that

File: src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        byte[] buffer = new byte[BufferSize];
+        while (true)
+        {
+            int bytesRead = await contentStream.ReadAsync(buffer, cancellationToken);
+            if (bytesRead == 0)
+                break;
+            await snapshotFileStream.WriteAsync(buffer.AsMemory(0, bytesRead), cancellationToken);
+
+            progressTracker.AddProgress(bytesRead);
+        }
```
Comment (emlautarom1) [inline src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs:122]: According to https://stackoverflow.com/a/230141/11035317 this is not the fastest way of doing it. Is this something we should focus on?

Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs:122]: the problem with the CopyTo method is that it doesn't allow me to track progress. Maybe we could wrap the contentStream so we can count read bytes 🤔

**Review by Demuirgos — APPROVED** [review]

File: src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs (hunk 1 above)
Comment (Demuirgos) [reply src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs:122]: I think [what they do in ``CopyToAsync``](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs,87) is similar to what is done here already

Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs:122]: yes, I was looking into it and I think the same

File: src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            snapshotConfig.Checksum is null ? null : Bytes.FromHexString(snapshotConfig.Checksum);
+
+        string snapshotFileName = Path.GetTempFileName();
+
+        await DownloadSnapshotTo(snapshotUrl, snapshotFileName, cancellationToken);
+
+        if (snapshotChecksum is not null)
+        {
+            bool isChecksumValid = await VerifyChecksum(snapshotFileName, snapshotChecksum);
+            if (!isChecksumValid)
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs:66]: NP: Should it delete the file?

**Review by asdacap — APPROVED** [review]

File: src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs (hunk 2 above)
Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs:76]: done

File: src/Nethermind/Nethermind.Init.Snapshot/ProgressTracker.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+        _timer = timerFactory.CreateTimer(interval);
+        _timer.Elapsed += ReportProgress;
+        _timer.AutoReset = false;
+    }
+
+    public void AddProgress(long count)
+    {
+        _current += count;
+        _timer.Enabled = true;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Init.Snapshot/ProgressTracker.cs:33]: if `AutoReset` is false you probably don't need to set true all the time.

**Comment by rubo** [comment]
Are we sure we don't need tests at all? Isn't it possible to simulate a download?

**Review by LukaszRozmej — COMMENTED** [review]
You could probably combine `ZipArchive`, and custom Stream to have:
- Downloading
- Extracting
- Checksum check
- Progress tracking

all in one go, directly from the wire

**Comment by jmederosalvarado** [comment]
@LukaszRozmej downloading and checksum verification can be done in one go directly from the wire, extracting needs seeking back and forth, so it's not possible to do everything in one pass.

**Pull request merged**